go/consensus/tendermint/apps/beacon/backend_vrf.go

```diff
@@ -0,0 +1,488 @@
+	entropy := GetBeacon(future.Epoch, prodEntropyCtx, req.Header.GetLastCommitHash())
+	if err = impl.app.onNewBeacon(ctx, entropy); err != nil {
+		return fmt.Errorf("beacon: failed to generate debug entropy")
+	}
+
+	// Emit the new VRF alpha event.
+	ctx.EmitEvent(api.NewEventBuilder(impl.app.Name()).TypedAttribute(&beacon.VRFEvent{
+		Epoch:       vrfState.Epoch,
+		Alpha:       vrfState.Alpha,
+		SubmitAfter: vrfState.SubmitAfter,
+	}))
+
+	return nil
+}
+
+func (impl *backendVRF) ExecuteTx(
+	ctx *api.Context,
+	state *beaconState.MutableState,
+	params *beacon.ConsensusParameters,
+	tx *transaction.Transaction,
+) error {
+	switch tx.Method {
+	case beacon.MethodVRFProve:
+		return impl.doProveTx(ctx, state, params, tx)
+	case MethodSetEpoch:
+		if !params.DebugMockBackend {
+			return fmt.Errorf("beacon: method '%s' is disabled via consensus", MethodSetEpoch)
+		}
+		return impl.doSetEpochTx(ctx, state, tx.Body)
+	default:
+		return fmt.Errorf("beacon: invalid method: %s", tx.Method)
+	}
+}
+
+func (impl *backendVRF) doProveTx(
+	ctx *api.Context,
+	state *beaconState.MutableState,
+	params *beacon.ConsensusParameters,
+	tx *transaction.Transaction,
+) error {
+	vrfState, err := state.VRFState(ctx)
+	if err != nil {
+		return fmt.Errorf("beacon: failed to get VRF state: %w", err)
+	}
+	if vrfState == nil {
+		return fmt.Errorf("beacon: no VRF state")
+	}
+
+	// Ensure that the minimum delay has passed since alpha was generated.
+	if ctx.BlockHeight()+1 <= vrfState.SubmitAfter {
+		return fmt.Errorf("beacon: premature VRF proof")
+	}
+
+	// Ensure the tx is from a current valid participant.
+	registryState := registryState.NewMutableState(ctx.State())
+	node, err := registryState.Node(ctx, ctx.TxSigner())
+	if err != nil {
+		return fmt.Errorf("beacon: tx not from a node: %v", err)
+	}
+	if node.VRF == nil {
+		return fmt.Errorf("beacon: tx signer missing VRF metadata")
+	}
+
+	// Deserialize the tx.
+	var proveTx beacon.VRFProve
+	if err = cbor.Unmarshal(tx.Body, &proveTx); err != nil {
+		return fmt.Errorf("beacon: failed to deserialize prove tx: %w", err)
+	}
+	if proveTx.Epoch != vrfState.Epoch {
+		return fmt.Errorf("beacon: proof for invalid epoch: %d", proveTx.Epoch)
+	}
+
+	// Verify the proof.
+	proof := signature.Proof{
+		PublicKey: node.VRF.ID,
+	}
+	if err = proof.Proof.UnmarshalBinary(proveTx.Pi); err != nil {
+		return fmt.Errorf("beacon: failed to deserialize raw proof: %w", err)
+	}
+	ok, beta := proof.Verify(vrfState.Alpha)
+	if !ok {
+		return fmt.Errorf("beacon: failed to verify beta")
+	}
+
+	if oldPi := vrfState.Pi[node.ID]; oldPi == nil {
+		// Fresh proof, store pi.
+		vrfState.Pi[node.ID] = &proof
+		if err = state.SetVRFState(ctx, vrfState); err != nil {
+			return fmt.Errorf("beacon: failed to update state: %w", err)
+		}
+	} else {
+		// That's odd, the node already submitted a proof, ensure that the
+		// betas match, and if not, reject and consider slashing.
```

Comment:
The failure case is impossible assuming the VRF is specified/implemented correctly since the beta output is deterministic.  I could have the byzantine node re-submit the tx with the same proof though.

Reply:
This actually is kind of hard to test, because when using mock timekeeping the backend will immediately transition the epoch once it receives all outstanding VRF proofs, if a transition is queued.  Since I do not want to mess with timekeeping for test cases any more (and not having this workaround makes things extremely slow), I am going to punt on this for now because the logic is trivially correct.